packages/grant-explorer/src/store.ts

```diff
@@ -46,19 +46,49 @@ const defaultVotingTokens = Object.fromEntries(
   })
 ) as Record<ChainId, VotingToken>;
 
+function areProjectIdsEqual(a: CartProject, b: CartProject): boolean {
+  return (
+    a.grantApplicationId === b.grantApplicationId && a.chainId === b.chainId
+  );
+}
+
+function updateOrInsertCartProject(
+  currentProjects: CartProject[],
+  newProject: CartProject
+): CartProject[] {
+  const initialAcc: {
+    projects: CartProject[];
+    hasUpdatedProject: boolean;
+  } = {
+    projects: [],
+    hasUpdatedProject: false,
+  };
+
+  const result = currentProjects.reduce((acc, project) => {
+    if (areProjectIdsEqual(project, project)) {
+      return {
+        projects: [...acc.projects, newProject],
+        hasUpdatedProject: true,
+      };
+    } else {
+      return { ...acc, projects: [...acc.projects, project] };
+    }
+  }, initialAcc);
+
+  return result.hasUpdatedProject
+    ? result.projects
+    : [...currentProjects, newProject];
+}
```

Comment:
I'm not sure I interpret the above correctly. Could it be done like this?

```ts
function updateOrInsertCartProject(
  currentProjects: CartProject[],
  newProject: CartProject
): CartProject[] {
  if (currentProjects.find(p => p.id === newProject.id) !== undefined) {
    return currentProjects;
  } else {
    return [...currentProjects, newProject];
  }
}
```

Reply:
this snippet doesn't update the project when it already exists, it only appends

Reply:
I thought a project is a data object? Does it have anything mutable?